qcodes/instrument/parameter.py

```diff
@@ -434,17 +415,14 @@ def snapshot_base(self, update: bool = True,
                 f"Parameter ({self.name}) is used in the snapshot while it "
                 f"should be excluded from the snapshot")
 
-        if hasattr(self, 'get') and self._snapshot_get \
-                and self._snapshot_value and update:
-            self.get()
+        state: Dict[str, Any] = {'ts': self.cache.timestamp,
+                                 '__class__': full_class(self),
+                                 'full_name': str(self)}
 
-        state: Dict[str, Any] = copy(self._latest)
-        state['__class__'] = full_class(self)
-        state['full_name'] = str(self)
-
-        if not self._snapshot_value:
-            state.pop('value')
-            state.pop('raw_value', None)
+        if self._snapshot_value:
+            get_if_invalid = self._snapshot_get and update
```

Comment:
* If update is true this should explicitly call parameter.get as update means don't trust the cache IMHO
* Do we need to keep the has get check?


So should it be?
```python
has_get = hasattr(self, 'get')
if self._snapshot_value: 
    if update:
         state['value'] = self.get()
    else:
        get_if_invalid = self._snapshot_get and has_get:
        state['value'] = self.cache.get(get_if_invalid=get_if_invalid)
```

Reply:
@jenshnielsen I thought about this suggestion and I do agree with it!

With one minor modification - [as per docstring of "snapshot_get" flag](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument/parameter.py#L168-L171), it prevents calling get even `update=True`. Also, i think it's good that if parameter is not gettable even with `update=True` its get doesnt get called to avoid useless exception.

Hence i propose the following (also took the liberty to try to name variables in a meaningful way):
```python
if self._snapshot_value:
    has_get = hasattr(self, 'get')
    allowed_to_call_get_when_snapshotting = self._snapshot_get
    can_call_get_when_snapshotting = allowed_to_call_get_when_snapshotting and has_get
    if can_call_get_when_snapshotting and update:
        state['value'] = self.get()
    else:
        state['value'] = self.cache.get(get_if_invalid=can_call_get_when_snapshotting)
    state['raw_value'] = self.cache.raw_value
```
what do you think?

Reply:
Yes that looks right

Reply:
great! i'm adding this to the code and covering it with tests...